src/nrniv/nrnapi.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include "hocdec.h"
+#include "section.h"
+
+// we define these here to allow the API to be independent of internal details
+#define STACK_IS_STR 1
+#define STACK_IS_VAR 2
+#define STACK_IS_NUM 3
+#define STACK_IS_OBJVAR 4
+#define STACK_IS_OBJTMP 5
+#define STACK_IS_USERINT 6
+#define STACK_IS_SYM 7
+#define STACK_IS_OBJUNREF 8
+#define STACK_UNKNOWN -1
+
+class SectionListIterator {
+public:
+  SectionListIterator(hoc_Item *);
+  Section *next(void);
+  int done(void);
+
+private:
+  hoc_Item *initial;
+  hoc_Item *current;
+};
+
+class SymbolTableIterator {
+public:
+  SymbolTableIterator(Symlist *);
+  char const *next(void);
+  int done(void);
+
+private:
+  Symbol *current;
+};
+
+extern "C" {
+/****************************************
+ * Initialization
+ ****************************************/
+int nrn_init(int argc, const char **argv);
+void nrn_redirect_stdout(int (*myprint)(int, char *));
+
+/****************************************
+ * Sections
+ ****************************************/
+Section *nrn_new_section(char const *const name);
+void nrn_connect_sections(Section *child_sec, double child_x,
+                          Section *parent_sec, double parent_x);
+void nrn_set_section_length(Section *sec, double length);
+double nrn_get_section_length(Section *sec);
+double nrn_get_section_Ra(Section *sec);
+void nrn_set_section_Ra(Section *sec, double val);
+char const *nrn_secname(Section *sec);
+void nrn_push_section(Section *sec);
+void nrn_pop_section(void);
+void nrn_insert_mechanism(Section *sec, Symbol *mechanism);
+hoc_Item *nrn_get_allsec(void);
+hoc_Item *nrn_get_sectionlist_data(Object *obj);
+
+/****************************************
+ * Segments
+ ****************************************/
+int nrn_get_nseg(Section const *const sec);
+void nrn_set_nseg(Section *const sec, const int nseg);
+void nrn_set_segment_diam(Section *const sec, const double x,
+                          const double diam);
+double *nrn_get_rangevar_ptr(Section *const sec, Symbol *const sym,
+                             double const x);
+
+/****************************************
+ * Functions, objects, and the stack
+ ****************************************/
+Symbol *nrn_get_symbol(char const *const name);
+int nrn_get_symbol_type(Symbol *sym);
+// TODO: need a way of mapping type identifiers to meaningful names
+double *nrn_get_symbol_ptr(Symbol *sym);
+void nrn_push_double(double val);
+double nrn_pop_double(void);
+void nrn_push_double_ptr(double *addr);
+double *nrn_pop_double_ptr(void);
```

Comment:
Note that with #2027 there will be a new handle-to-double type (`data_handle<double>`, but it's C++ so would have to be wrapped in a C-compatible API) that must be used to refer to model data in place of raw pointers if the reference needs to remain valid after operations that modify the model.

It might be better to preemptively expose a more generic API than "double pointer".

Also, these methods should come with explicit documentation about the ownership semantics of the returned value. Who is responsible for freeing the pointed-to `double`?

Reply:
Situation gets improved in the follow-up PR